### PR TITLE
[Executor][Internal]Improve inputs mapping's error message.

### PR DESCRIPTION
**Pull request by huaiyan**
# Description

Improve inputs mapping's error message.
Use message_format and improve user error's error message to guide customer to fix the issue

# All Promptflow Contribution checklist:
- [X] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 1]
```diff
@@ -793,7 +796,9 @@ def _stringify_generator_output(self, outputs: dict):
 
     def _submit_to_scheduler(self, context: FlowExecutionContext, inputs, nodes: List[Node]) -> Tuple[dict, dict]:
         if not isinstance(self._node_concurrency, int):
-            raise NodeConcurrencyNotFound("Need to set node concurrency as using flow executor.")
+            raise NodeConcurrencyNotFound(
+                message_format=f"Need to set valid node concurrency to run, current value is {self._node_concurrency}."
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:800]: Is it good to format the {self._node_concurrency} into the message?

Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:800]: How does the user handle this error? We should provide a troubleshooting guide in the error message.
i.e. We can tell the user what are valid node concurrency values.

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 2]
```diff
@@ -854,11 +859,14 @@ def apply_inputs_mapping(
...
-                f"Couldn't find these mapping relations: {', '.join(notfound_mapping_relations)}. "
+            invalid_relations = ", ".join(notfound_mapping_relations)
+            raise InputMappingError(
+                message_format="The input for flow is incorrect. "
+                "Couldn't find these mapping relations: {invalid_relations}. "
                 "Please make sure your input mapping keys and values match your YAML input section and input data. "
-                "If a mapping value has a '${data' prefix, it might be generated from the YAML input section, "
-                "and you may need to manually assign input mapping based on your input data."
+                "If a mapping reads input from 'data', it might be generated from the YAML input section, "
+                "and you may need to manually assign input mapping based on your input data.",
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:868]: Let's add a `()` to multi-line messages.
Like this:
image

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 3]
```diff
@@ -951,11 +967,17 @@ def _apply_inputs_mapping_for_all_lines(
         }]
         """
         if inputs_mapping is None:
-            # This exception should not happen since we should use default inputs_mapping if not provided.
-            raise NoneInputsMappingIsNotSupported("Inputs mapping is None.")
+            # This exception should not happen since we will use default inputs_mapping for None input.
+            raise NoneInputsMappingIsNotSupported(
+                message_format="Inputs mapping is None. "
+                "Please set one inputs mapping or use default inputs mapping in flow_executor."
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:973]: This is a system error, right? Can the user understand this troubleshoot message?

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 4]
```diff
@@ -951,11 +967,17 @@ def _apply_inputs_mapping_for_all_lines(
...
+            )
         merged_list = FlowExecutor._merge_input_dicts_by_line(input_dict)
         if len(merged_list) == 0:
-            raise EmptyInputAfterMapping("Input data does not contain a complete line. Please check your input.")
+            raise InputMappingError(
+                message_format="The input for flow is incorrect. All inputs come from different line of data "
+                "so can't execute. Please provide data in same line to correct the input."
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:979]: I think this error message can be refined to be more native. We may ask for ChatGPT's help.

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 5]
```diff
@@ -870,8 +878,13 @@ def _merge_input_dicts_by_line(
         input_dict: Mapping[str, List[Mapping[str, Any]]],
     ) -> List[Mapping[str, Mapping[str, Any]]]:
         for input_key, list_of_one_input in input_dict.items():
-            if len(list_of_one_input) == 0:
-                raise EmptyInputListError(f"List from key '{input_key}' is empty.")
+            if not list_of_one_input:
+                raise InputMappingError(
+                    message_format="The input for flow is incorrect. Input from key '{input_key}' is one empty list. "
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:883]: one -> an?

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 6]
```diff
@@ -870,8 +878,13 @@ def _merge_input_dicts_by_line(
...
+                    "Which means we cannot generate one line for the flow run. "
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:884]: "Which" is not a complete sentence. We should merge this part with the previous one. Something like:
"Input from key '{input_key}' is an empty list, which means ..."

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 3 above)
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:973]: The troubleshoot message is intended to be read by partner team such as PRS.

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 1 above)
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:800]: Usually it's set by `exec_bulk`, `exec_line`.
The content will be changed, but we may need it to investigate.
Although I'm not sure its content will never be customer content.

Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:800]: It's one System error. Customer usually can't do anything for it.

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 7]
```diff
@@ -28,12 +28,15 @@
...
-from promptflow.exceptions import ErrorTarget, PromptflowException, SystemErrorException, ValidationException
+from promptflow.exceptions import PromptflowException
 from promptflow.executor import _input_assignment_parser
 from promptflow.executor._errors import (
+    InputMappingError,
     InvalidAggregationInput,
     NodeConcurrencyNotFound,
     NodeOutputNotFound,
+    NodeResultCountNotMatch,
+    NoneInputsMappingIsNotSupported,
```
[the remark this answers is not in the document]
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:39]: Offline synced with others.
We will introduce one UnexpectedError in _core._errors as one common SystemError for this kind of case.
(very sparse case and hard to find one good name)

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 8]
```diff
@@ -880,9 +893,12 @@ def _merge_input_dicts_by_line(
             if not any(LINE_NUMBER_KEY in one_item for one_item in list_of_one_input)
         }
         if len(set(all_lengths_without_line_number.values())) > 1:
-            raise LineNumberNotAlign(
+            raise InputMappingError(
+                message_format="The input for flow is incorrect. "
                 "Line numbers are not aligned. Some lists have dictionaries missing the 'line_number' key, "
-                f"and the lengths of these lists are different. List lengths: {all_lengths_without_line_number}"
+                "and the lengths of these lists are different. List lengths are: {all_lengths_without_line_number}. "
+                "Please make sure these lists have the same length or add 'line_number' key to each dictionary.",
```
[the remark this answers is not in the document]
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:900]: `all_lengths_without_line_number` is one dict.
And we may don't want to include customer input key in `message_format`.

File: src/promptflow/tests/executor/unittests/executor/test_flow_executor.py [hunk 9]
```diff
@@ -66,11 +63,11 @@ def test_apply_inputs_mapping(self, inputs, inputs_mapping, expected):
...
                     "answer": "${data.output}",
                 },
-                MappingSourceNotFound,
-                "Couldn't find these mapping relations: ${baseline.output}, ${data.output}. "
-                "Please make sure your input mapping keys and values match your YAML input section and input data. "
-                "If a mapping value has a '${data' prefix, it might be generated from the YAML input section, "
-                "and you may need to manually assign input mapping based on your input data.",
+                InputMappingError,
+                "The input for flow is incorrect. Couldn't find these mapping relations: ${baseline.output}, "
+                "${data.output}. Please make sure your input mapping keys and values match your YAML input section "
```
[the remark this answers is not in the document]
Comment (huaiyan) [reply src/promptflow/tests/executor/unittests/executor/test_flow_executor.py:68]: No, for portal experience, customer must fill in input mapping for each input. So we will not use default mapping relation.

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 4 above)
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:979]: @chw-microsoft 
Do we have template for system error's message about asking support for help?
Below is the most similar message for me.
image

File: src/promptflow/tests/executor/unittests/executor/test_flow_executor.py [hunk 10]
```diff
@@ -303,7 +303,7 @@ def test_validate_and_apply_inputs_mapping_error(self, inputs_mapping, error_cod
                 },
                 None,
                 NoneInputsMappingIsNotSupported,
-                "Inputs mapping is None.",
+                "Inputs mapping is None. Please set one inputs mapping or use default inputs mapping in flow_executor.",
```
[the remark this answers is not in the document]
Comment (huaiyan) [reply src/promptflow/tests/executor/unittests/executor/test_flow_executor.py:306]: This is one SystemError, and it's one hint intend to be read by partner team such as PRS.

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 5 above)
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:883]: Fixed.

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 6 above)
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:884]: Fixed.

File: src/promptflow/tests/executor/unittests/executor/test_flow_executor.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (huaiyan) [reply src/promptflow/tests/executor/unittests/executor/test_flow_executor.py]: Offline synced. We should not use error_message as hint for partner team to fix issue.

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 2 above)
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py]: Fixed.

File: src/promptflow/promptflow/_core/_errors.py [hunk 11]
```diff
@@ -4,6 +4,12 @@
 from promptflow.exceptions import ErrorTarget, SystemErrorException, UserErrorException, ValidationException
 
 
+class UnexpectedError(SystemErrorException):
+    """Exception raised for normal errors that should not happen."""
```
Comment (chw-microsoft) [inline src/promptflow/promptflow/_core/_errors.py:8]: Exception raised for unexpected errors that should not occur under normal circumstances.

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 12]
```diff
@@ -244,7 +244,9 @@ def load_and_exec_node(
             node_runs = run_tracker.collect_node_runs()
             if len(node_runs) != 1:
                 # Should not happen except there is bug in run_tracker or thread control.
-                raise NodeResultCountNotMatch("Expecting eactly one node run.")
+                raise UnexpectedError(
+                    message_format="Expecting only one node result. Please contact support for further assistance."
+                )
```
Comment (chw-microsoft) [inline src/promptflow/promptflow/executor/flow_executor.py:249]: To be informative:  Single node execution failed. Expected one node result, but received {node_result_num}. Please contact support for further assistance.

I add template as well in sheet

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 13]
```diff
@@ -793,15 +795,21 @@ def _stringify_generator_output(self, outputs: dict):
 
     def _submit_to_scheduler(self, context: FlowExecutionContext, inputs, nodes: List[Node]) -> Tuple[dict, dict]:
         if not isinstance(self._node_concurrency, int):
-            raise NodeConcurrencyNotFound("Need to set node concurrency as using flow executor.")
+            raise UnexpectedError(
+                message_format=(
+                    "Need to set valid node concurrency to run, current value is {current_value}. "
+                    "Please contact support for further assistance."
```
Comment (chw-microsoft) [inline src/promptflow/promptflow/executor/flow_executor.py:801]: Flow execution failed. To proceed, ensure that a valid node concurrency value is set; the current value is {current_value}. Please contact support for further assistance.

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 14]
```diff
@@ -854,11 +862,15 @@ def apply_inputs_mapping(
...
-                f"Couldn't find these mapping relations: {', '.join(notfound_mapping_relations)}. "
-                "Please make sure your input mapping keys and values match your YAML input section and input data. "
-                "If a mapping value has a '${data' prefix, it might be generated from the YAML input section, "
-                "and you may need to manually assign input mapping based on your input data."
+            invalid_relations = ", ".join(notfound_mapping_relations)
+            raise InputMappingError(
+                message_format=(
+                    "The input for flow is incorrect. Couldn't find these mapping relations: {invalid_relations}. "
+                    "Please make sure your input mapping keys and values match your YAML input section and input data. "
+                    "If a mapping reads input from 'data', it might be generated from the YAML input section, "
```
Comment (chw-microsoft) [inline src/promptflow/promptflow/executor/flow_executor.py:870]: I might be confused with this expression.

_If a mapping reads input from 'data'_

Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:870]: Yes... The previous version used "${data" but met exception as formatting the error message.

Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:875]: Add todo to replace the detail message about default mapping by doc link

**Review by chw-microsoft — APPROVED** [review]

File: src/promptflow/promptflow/_core/_errors.py (hunk 11 above)
Comment (huaiyan) [reply src/promptflow/promptflow/_core/_errors.py]: Great, thanks~

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 12 above)
Comment (huaiyan) [reply src/promptflow/promptflow/executor/flow_executor.py:253]: Good catch.

**Pull request merged**